### PR TITLE
OpenMP Groupby any/all

### PR DESCRIPTION
**Pull request by WillAyd**
Seems to give a small boost with our current test suite, though this could depend on the shape of the input data:

```sh
       before           after         ratio
     [824fc830]       [f7d33366]
     <main>           <parallel-shift>
-      46.3±0.4μs       41.6±0.4μs     0.90  groupby.GroupByMethods.time_dtype_as_group('datetime', 'any', 'direct', 5)
-     1.26±0.01ms      1.13±0.01ms     0.89  groupby.GroupByMethods.time_dtype_as_field('object', 'any', 'transformation', 5)
-     1.21±0.05ms      1.07±0.02ms     0.89  groupby.GroupByMethods.time_dtype_as_field('int', 'any', 'transformation', 5)
```

For the openmp link arg meson should have a much better facility for auto-detecting that which we can leverage

Cython dev docs have been updated with more info on parallelization:

https://github.com/cython/cython/blob/master/docs/src/tutorial/parallelization.rst

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I tried this recently ish with libalgos.nancorr and couldn’t get it working on Mac. Would be awesome if you can figure it out!

**Comment by WillAyd** [comment]
Assuming this is true looks like the macOS bundled llvm just doesn't support llvm

https://stackoverflow.com/a/60043467/621736

Meson should help us compared to setuptools - will rebase after that goes into main and see

File: meson.build [hunk 1]
```diff
@@ -22,7 +22,7 @@ py = py_mod.find_installation('python')
 py_dep = py.dependency()
 tempita = files('generate_pxi.py')
 versioneer = files('generate_version.py')
-
+openmp = dependency('openmp', required: false)
```
Comment (jbrockmendel) [inline meson.build:25]: does this mean that i shouldn't expect this PR to affect perf on mac?

Comment (WillAyd) [reply meson.build:25]: If using the standard compiler I don't think so, but you may be able to achieve it if you use a conda supplied installer. 

There are two ways to validate:
  1. When installing with meson add the `--config-settings=compile-args="--verbose"` flag, i.e. `python -m pip install -ve . --no-build-isolation --config-settings=compile-args="--verbose"` and search for the `-fopenmp` flag being passed when linking the groupby lib
  2. grep the groupby lib for any openmp symbols. On linux this would be `objdump --syms build/cp311/pandas/_libs/groupby.cpython-311-x86_64-linux-gnu.so | grep omp_get` with the standard meson build directory + py311. You will need to tweak to your environment, and maybe tweak objdump (not sure how it works on macOS)

Here's the sample output from 2 on my computer:

```sh
objdump --syms build/cp311/pandas/_libs/groupby.cpython-311-x86_64-linux-gnu.so | grep omp_get
0000000000000000       F *UND*	0000000000000000              omp_get_thread_num@OMP_1.0
0000000000000000       F *UND*	0000000000000000              omp_get_num_threads@OMP_1.0
```

The `omp_get` symbols are of interest

Comment (WillAyd) [reply meson.build:25]: I guess the second solution above is the only "guaranteed" way of knowing; for option 1 it is possible for the flag to be silently ignored

File: pandas/_libs/groupby.pyx [hunk 2]
```diff
@@ -621,7 +622,7 @@ def group_any_all(
     out[:] = 1 - flag_val
 
     with nogil:
-        for i in range(N):
+        for i in prange(N):
```
Comment (jbrockmendel) [inline pandas/_libs/groupby.pyx:625]: any particular reason this went here rather than someplace else?  my assumption is it was a pretty arbitrary choice made for a POC, but worth checking.

if this goes in, can i look forward to a bunch more PRs adding prange all over the place?

Comment (WillAyd) [reply pandas/_libs/groupby.pyx:625]: To add some context to the above, one of the first algorithms I looked at was corr since our implementation lags pretty bad relative to the NumPy implementation (granted we handle NA values and they do not)

https://github.com/pandas-dev/pandas/blob/94e868a76f5dfd30e84469b47a316699eb1f083d/pandas/_libs/algos.pyx#L365

Here is what I tried to change:

```python
    with nogil:
        for xi in prange(K):
            for yi in range(xi + 1):
                # Welford's method for the variance-calculation
                # https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
                nobs = ssqdmx = ssqdmy = covxy = meanx = meany = 0
                for i in range(N):
                    if mask[i, xi] and mask[i, yi]:
                        vx = mat[i, xi]
                        vy = mat[i, yi]
                        nobs += 1
                        dx = vx - meanx
                        dy = vy - meany
                        meanx += 1. / nobs * dx
                        meany += 1. / nobs * dy
                        ssqdmx += (vx - meanx) * dx
                        ssqdmy += (vy - meany) * dy
                        covxy += (vx - meanx) * dy

                if nobs < minpv:
                    result[xi, yi] = result[yi, xi] = NaN
                else:
                    divisor = (nobs - 1.0) if cov else sqrt(ssqdmx * ssqdmy)

                    if divisor != 0:
                        result[xi, yi] = result[yi, xi] = covxy / divisor
                    else:
                        result[xi, yi] = result[yi, xi] = NaN
```

And Cython gives this error when you try to generate it:

```sh
  Error compiling Cython file:
  ------------------------------------------------------------
  ...
                  for i in range(N):
                      if mask[i, xi] and mask[i, yi]:
                          vx = mat[i, xi]
                          vy = mat[i, yi]
                          nobs += 1
                          dx = vx - meanx
                                   ^
  ------------------------------------------------------------

  /home/willayd/clones/pandas/pandas/_libs/algos.pyx:377:34: Cannot read reduction variable in loop body
```

So I think all those values like `meanx`, `meany`, etc.. that accumulate a value over a loop end up with pretty limited value, even if that accumulation should happen all on one thread. My guess is Cython is being ultra-conservative about this, so as we learn more about it can definitely give feedback upstream and see what they say

**Review by jreback — CHANGES_REQUESTED** [review]
do we really think ipenmp is a good idea?
this opens up a huge box

**Comment by WillAyd** [comment]
Not sure. It definitely is the easiest way for us to parallelize things via Cython (where applicable), but I can't claim to have experience with using it otherwise. 

The packaging seems like the riskiest piece. Maybe we have to turn it off for wheels but enable for conda environments where openmp can be ensured?

Are there other huge downsides anyone is aware of?

File: pandas/_libs/groupby.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby.pyx:625]: Makes sense, thanks for taking a look.  I'll poke at that one if/when this goes in, as it definitely seems like it should parallelize nicely.

**Comment by rhshadrach** [comment]
In general, you don't want to parallelize at multiple levels of your code. E.g. a workflow that uses up all cores via multiprocessing a workflow and then trying to use all cores within each process will suffer from context switches.

Another case is in a large shared server were you want to ensure you aren't over utilizing resources.

**Comment by h-vetinari** [comment]
> @h-vetinari wondering if you have any experience with distributing openmp linked libraries via conda and pip. Not sure how the presence of this on the build system may affect downstream users

TL;DR: It works fine with conda, not fine with pip.

In more detail, conda-forge has all the required bits & pieces to make this work (and we [ship](https://github.com/conda-forge/openblas-feedstock/blob/main/recipe/meta.yaml#L27) openmp-support in openblas for example), though windows has the wrinkle that the `intel-openmp` and `llvm-openmp` implementations are different and cannot be co-installed. We've been wanting to switch to `llvm-openmp` also on windows (see e.g. [here](https://github.com/conda-forge/openblas-feedstock/pull/115), but that's been blocked on some Fortran infrastructure work that's still pending).

As a concrete example, scipy has [not allowed](https://github.com/scipy/scipy/issues/10239#issuecomment-795030817) a hard dependence on OpenMP for various reasons, primarily related to the limitations of PyPI's distribution model. @rgommers was & is heavily involved in this, and also did a [write-up](https://pypackaging-native.github.io/key-issues/native-dependencies/blas_openmp/) about this very subject in the context of the pypackaging-native initiative. From a recently-added cross-reference on that scipy-issue, I see that a bunch of familiar faces already [discussed](https://github.com/Quansight-Labs/nasa-roses-2020/issues/27) this in the wider DataPy context, including pandas.

**Comment by mroeschke** [comment]
FWIW numba seems to "select a threading layer if already available" instead of requiring one https://numba.pydata.org/numba-doc/latest/user/threading-layer.html#selecting-a-named-threading-layer

> you don't want to parallelize at multiple levels of your code

Yeah numba specifically detects and disallows this when jitting which might be hard to do on our side

**Comment by WillAyd** [comment]
> As a concrete example, scipy has [not allowed](https://github.com/scipy/scipy/issues/10239#issuecomment-795030817) a hard dependence on OpenMP for various reasons, primarily related to the limitations of PyPI's distribution model. @rgommers was & is heavily involved in this, and also did a [write-up](https://pypackaging-native.github.io/key-issues/native-dependencies/blas_openmp/) about this very subject in the context of the pypackaging-native initiative. From a (somewhat-)recently-added cross-reference on that scipy-issue, I see that a bunch of familiar faces already [discussed](https://github.com/Quansight-Labs/nasa-roses-2020/issues/27) this in the wider PyData context, including pandas.

Wow thanks these are great resources. But I think the last link might be private? I get a 404 trying to access

**Comment by h-vetinari** [comment]
> But I think the last link might be private? I get a 404 trying to access

My bad, sorry, I'm not used to being a member of non-public GH orgs 😅

**Comment by jreback** [comment]
I am -1 on randomly adding parallelizable that is 

- not external controllable (via keyword and env variable)
- requires non standard things (openmp is standard but not generally available)
- defaults to anything hut False

this needs comprehensive policies otherwise you end up with multiple of layers of parallelism which is almost always bad

point is that this needs substantial testing and validation and tbh we may want to punt completely at the pandas level on this

**Comment by jbrockmendel** [comment]
I'm open to a "needs more thought" argument, but on a lot of fronts we've done a pretty good job with optimization so performance gains are going to _have_ to come from parallelism

**Comment by WillAyd** [comment]
Yea sounds like generally applying has some caveats. A conservative approach to start could be to add `-Duse-omp=ON` as a meson option, but set the default to false. That would give people building from source an easy way to opt in, if they don't plan on using pandas in a multiprocessing environment

**Comment by rgommers** [comment]
+1 to @jreback's points. I'll note that @thomasjpfan put quite some thought into this topic and wrote up a design proposal for enabling parallelism across core PyData packages in a careful fashion: https://thomasjpfan.github.io/parallelism-python-libraries-design/ (also going to be a SciPy'23 talk I believe).

**Comment by WillAyd** [comment]
Let's close this - I don't plan on taking any more immediate action. I think this is a good reference for any future discussion of the topic

**Comment by thomasjpfan** [comment]
Looking over the concerns from this PR, I can see the following path forward for pandas + OpenMP:

1. Let the default be **single threaded**.
2. Pass a keyword argument from the Python layer all the way down to `Cython`'s prange. This makes it harder to engage in nested parallelism in pandas code.
3. Have OpenMP be configurable with a compile time flag. This allows Pyodide to build with OpenMP off and conda to build with it on. As noted in https://github.com/pandas-dev/pandas/pull/53149#issuecomment-1541009510, conda has a better story around packaging OpenMP.
4. Be ready for issues. Even without nested parallelism, there will be performance issues when there are two separate threadpools. Pandas will need to have some documentation warning users about configuring threads >= 2.
5. If pandas wants OpenMP support in the near-term, be prepared to ship a PyPI wheel with an OpenMP binary.

An alternative to OpenMP is to go with SciPy's solution, which developed their own [pthreads based solution](https://github.com/scipy/scipy/blob/762a349b62e59316617252c1d50371bab6fbee19/scipy/fft/_pocketfft/pocketfft_hdronly.h). It has the benefit of being a standard and avoids some of the issues with OpenMP.

**Comment by jbrockmendel** [comment]
@thomasjpfan can you comment on the pros/cons of the scipy solution, particularly for those of us with limited c++ experience

**Comment by thomasjpfan** [comment]
For reference, SciPy has discussed the topic of adding OpenMP and decided on pthreads: https://github.com/scipy/scipy/issues/10239 At a high level the pros and cons unique to SciPy's pthreads solution:

#### Pros
- Portable: pthreads is a POSIX standard
- None of the OpenMP headaches when libraries load different versions of OpenMP
- Easier to build. (No need to vendor OpenMP for wheels)

#### Cons
- As seen in SciPy's solution, pthreads needs a little more work to get going. OpenMP has higher level abstractions, which makes it easier to work with.
- Need to maintain a custom solution. (Although, if many libraries want to go down this route, we can wrap a C/C++ solution with a nice Cython frontend)
- Depending on how one implements the custom solution, you can still end up with CPU contention: https://github.com/xianyi/OpenBLAS/issues/4033
